src/Stack/QuerySplit.php

```diff
@@ -0,0 +1,110 @@
+
+namespace Stack;
+
+/**
+ * This plugin is used together with ProxySQL and has the role of annotating queries as write in
+ * special situations.
+ *
+ * The plugins hooks on Wordpress "query" hook and process the query and if necessary the read
+ * queries are send to master.
+ */
+
+/**
+ * Class QuerySplit
+ */
+class QuerySplit
+{
+
+    const MASTER_LABEL = "/*master*/";
+    const SLAVE_LABEL = "/*slave*/";
+
+    private $srtm = false;
+
+    public function __construct()
+    {
+        $this->presetup();
+        $this->addFilters();
+    }
+
+    private function addFilters()
+    {
+        add_filter('query', [$this, 'processQuery']);
+    }
+
+    public function processQuery(string $query)
+    {
+        // don't need to analize the query if it's already write
+        // NOTE: this can be optimized to send reads on master per table
+        if (! $this->srtm === true) {
+            $this->analizeQuery($query);
+        }
+
+        // set the write label on the query
+        if ($this->srtm === true) {
+            $query = $query . " " . self::MASTER_LABEL;
+        }
+
+        return $query;
+    }
+
+    private function presetup()
+    {
+        // Send non-idempotent requests to master
+        if (! in_array($_SERVER['REQUEST_METHOD'], array( 'GET', 'HEAD' ))) {
+            $this->sendReadsToMaster();
+            return;
+        }
+
+        // Cron jobs always go to master so that they can run SQL queries more easily
+        if (isset($_GET['doing_wp_cron'])) {
+            $this->sendReadsToMaster();
+            return;
+        }
+
+        // TODO(@calind): why not sending to master all queries that are made from URI that starts with SITE_URL?
+        // admin-ajax.php uses master, is this intended behavior?
+
+        // Admin requests always go to master
+        if ('/wp/wp-admin/' == substr($_SERVER['REQUEST_URI'], 0, 13)) {
+            $this->sendReadsToMaster();
+            return;
+        }
+
+        if ('/wp/wp-login.php' == substr($_SERVER['REQUEST_URI'], 0, 16)) {
+            $this->sendReadsToMaster();
+            return;
+        }
+
+        if ('/wp/wp-register.php' == substr($_SERVER['REQUEST_URI'], 0, 19)) {
+            $this->sendReadsToMaster();
+            return;
+        }
+
+        // XML-RPC go to master since they usually are modifying content
+        if ('/wp/xmlrpc.php' == substr($_SERVER['REQUEST_URI'], 0, 14)) {
+            $this->sendReadsToMaster();
+            return;
+        }
+    }
+
+    public function sendReadsToMaster()
+    {
+        $this->srtm = true;
+    }
+
+    private function analizeQuery($query)
+    {
+        if ($this->isWriteQuery($query)) {
+            $this->sendReadsToMaster();
+        }
+    }
+
+    private function isWriteQuery($q)
+    {
+        // Quick and dirty: only SELECT statements are considered read-only.
+        $q = ltrim($q, "\r\n\t (");
+
+        return ! preg_match('/^(?:SELECT|SHOW|DESCRIBE|DESC|EXPLAIN)\s/i', $q);
```

Comment:
`SELECT ... FOR UPDATE` should go to master.

Reply:
`SELECT ... FOR UPDATE` is treated by the ProxySQL rules.

The `SHOW, DESCRIBE, DESC, EXPLAIN` will go to master because they are not labeled as force master.

This function is used to detect if a write (in the WordPress tables) happened during a PHP session and sends all other reads on the master to prevent read-after-write.

Here are the rules in ProxySQL:
```
mysql_query_rules =
(
        // some other wordpress specific queries that can be cached
        { rule_id = 1000, active = 1, digest = "0x8FA3D229330C516A", cache_ttl = 300000, apply = 1, comment = "Cache SELECT @@SESSION.sql_mode" },

	{rule_id=9000, active=1, match_pattern=".*\/\*master\*\/.*", destination_hostgroup=0, apply=1},
	{rule_id=9010, active=1, match_pattern=".*\/\*slave\*\/.*", destination_hostgroup=1, apply=1},

	// send reads to reader hostgroup
	{rule_id=9100, active=1, match_pattern="^SELECT.*FOR UPDATE$", destination_hostgroup=0, apply=1},
	{rule_id=9110, active=1, match_pattern="^SELECT.*", destination_hostgroup=1, apply=1},
)
```

Reply:
Still, queries made after `SELECT .. FOR UPDATE` should go to the master and this is mandatory, so it should set `srtm=true`.